urnai/sc2/states/terran_state.py

```diff
@@ -0,0 +1,81 @@
+import numpy as np
+from pysc2.env import sc2_env
+from pysc2.lib import units
+
+from urnai.constants import SC2Constants
+from urnai.sc2.states.states_utils import (
+    append_player_and_enemy_grids,
+    get_my_raw_units_amount,
+)
+from urnai.states.state_base import StateBase
+
+
+class TerranState(StateBase):
+
+    def __init__(
+        self,
+        grid_size: int = 4,
+        use_raw_units: bool = True,
+        raw_resolution: int = 64,
+    ):
+        self.grid_size = grid_size
+        self.player_race = sc2_env.Race.terran
+        self.use_raw_units = use_raw_units
+        self.raw_resolution = raw_resolution
+        self.reset()
+
+    def update(self, obs):
+        new_state = [
+            # Adds general information from the player.
+            obs.player.minerals / SC2Constants.MAX_MINERALS,
+            obs.player.vespene / SC2Constants.MAX_VESPENE,
+            obs.player.food_cap / SC2Constants.MAX_UNITS,
+            obs.player.food_used / SC2Constants.MAX_UNITS,
+            obs.player.food_army / SC2Constants.MAX_UNITS,
+            obs.player.food_workers / SC2Constants.MAX_UNITS,
+            (obs.player.food_cap - obs.player.food_used) / SC2Constants.MAX_UNITS,
+            obs.player.army_count / SC2Constants.MAX_UNITS,
+            obs.player.idle_worker_count / SC2Constants.MAX_UNITS,
+        ]
+
+        if self.use_raw_units:
+            new_state.extend(
+                [
+                    # Adds information related to player's Terran units/buildings.
+                    get_my_raw_units_amount(obs, units.Terran.CommandCenter)
+                    + get_my_raw_units_amount(obs, units.Terran.OrbitalCommand)
+                    + get_my_raw_units_amount(obs, units.Terran.PlanetaryFortress) / 2,
+                    get_my_raw_units_amount(obs, units.Terran.SupplyDepot) / 18,
+                    get_my_raw_units_amount(obs, units.Terran.Refinery) / 4,
+                    get_my_raw_units_amount(obs, units.Terran.EngineeringBay),
+                    get_my_raw_units_amount(obs, units.Terran.Armory),
+                    get_my_raw_units_amount(obs, units.Terran.MissileTurret) / 4,
+                    get_my_raw_units_amount(obs, units.Terran.SensorTower) / 1,
+                    get_my_raw_units_amount(obs, units.Terran.Bunker) / 4,
+                    get_my_raw_units_amount(obs, units.Terran.FusionCore),
+                    get_my_raw_units_amount(obs, units.Terran.GhostAcademy),
+                    get_my_raw_units_amount(obs, units.Terran.Barracks) / 3,
+                    get_my_raw_units_amount(obs, units.Terran.Factory) / 2,
+                    get_my_raw_units_amount(obs, units.Terran.Starport) / 2,
```

Comment:
This function goes through the same list several times just to perform a unit filter. Wouldn't it be better to turn this list into a dictionary using the unit type as the key? You can use `defaultdict` to return `0` in cases where there are no units of a given type.

Reply:
Do you mean creating a dictionary that maps the types present in `obs.raw_units` to the number of times it appears? If so, I believe the key needs to be a tuple of type, alliance and build_progress, since we need to count the number of units that are of some type, alliance and "complete" build progress.

Reply:
This is how I tried it:

```python
def default_value():
    return 0

def create_raw_units_amount_dict(
        obs : list,
        alliance : features.PlayerRelative = features.PlayerRelative.SELF
    ) -> defaultdict:
    dict = defaultdict(default_value)
    for unit in obs.raw_units:
        if (unit.alliance == alliance
        and unit.build_progress == SC2Constants.BUILD_PROGRESS_COMPLETE):
            dict[unit.unit_type] += 1
    return dict
```